edb/edgeql/compiler/inference/multiplicity.py

```diff
@@ -0,0 +1,701 @@
+"""
+
+
+from __future__ import annotations
+from typing import *
+
+import functools
+
+from edb import errors
+
+from edb.edgeql import qltypes
+
+from edb.schema import pointers as s_pointers
+from edb.schema import constraints as s_constr
+
+from edb.ir import ast as irast
+
+from . import cardinality
+from . import context as inference_context
+
+
+ZERO = qltypes.Multiplicity.ZERO
+ONE = qltypes.Multiplicity.ONE
+MANY = qltypes.Multiplicity.MANY
+
+
+def _max_multiplicity(
+    args: Iterable[qltypes.Multiplicity]
+) -> qltypes.Multiplicity:
+    # Coincidentally, the lexical order of multiplicity is opposite of
+    # order of multiplicity values.
+    arg_list = list(args)
+    if not arg_list:
+        return ZERO
+    else:
+        return min(arg_list)
+
+
+def _common_multiplicity(
+    args: Iterable[irast.Base],
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return _max_multiplicity(
+        infer_multiplicity(a, scope_tree=scope_tree, ctx=ctx) for a in args)
+
+
+@functools.singledispatch
+def _infer_multiplicity(
+    ir: irast.Expr,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # return MANY
+    raise ValueError(f'infer_multiplicity: cannot handle {ir!r}')
+
+
+@_infer_multiplicity.register
+def __infer_none(
+    ir: None,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # Here for debugging purposes.
+    raise ValueError('invalid infer_multiplicity(None, schema) call')
+
+
+@_infer_multiplicity.register
+def __infer_statement(
+    ir: irast.Statement,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return infer_multiplicity(
+        ir.expr, scope_tree=scope_tree, ctx=ctx)
+
+
+@_infer_multiplicity.register
+def __infer_empty_set(
+    ir: irast.EmptySet,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return ZERO
+
+
+@_infer_multiplicity.register
+def __infer_type_introspection(
+    ir: irast.TypeIntrospection,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # The result is always ONE, but we still want to actually
+    # introspect the expression.
+    return ONE
+
+
+def _infer_shape(
+    ir: irast.Set,
+    *,
+    is_mutation: bool=False,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> None:
+    for shape_set, _ in ir.shape:
+        new_scope = cardinality._get_set_scope(shape_set, scope_tree)
+        if shape_set.expr and shape_set.rptr:
+            expr_mult = infer_multiplicity(
+                shape_set.expr, scope_tree=new_scope, ctx=ctx)
+
+            ptrref = shape_set.rptr.ptrref
+            if (expr_mult is MANY and
+                    not ptrref.out_target.is_scalar and
```

Comment:
Use helpers from `irtypeutils.py` instead of accessing `TypeRef` members directly.

Reply:
Updated to use `is_object` instead.